blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+
+## Eviction warnings
+
+Actually this is not the first time I've thought of this issue. In 2014, I added [eviction warning][1] feature to sbt 0.13.6. In the lingo of dependency resolver, when you have two candidate versions 0.11.1 and 0.13.0, and when it picks 0.13.0, 0.11.1 is said to be evicted.
+
+~~~ text
+sbt:killer-app> evicted
+[warn] Found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
+[warn]  * io.circe:circe-core_2.12:0.13.0 is selected over 0.11.1
+[warn]      +- org.tpolecat:doobie-postgres-circe_2.12:0.10.0     (depends on 0.13.0)
+[warn]      +- de.heikoseeberger:akka-http-circe_2.12:1.26.0      (depends on 0.11.1)
+[success] Total time: 1 s, completed Feb 1, 2021, 2:59:56 PM
+~~~
+
+Here, we see that sbt successfully detected the conflict for circe-core.
+
+This was a step in the right direction, but it did not work well in reality because I had no way to tell whether two versions of a library would be binary compatible or not. In 2014 what I did was guess that a Java library would adopt [Semantic Versioning][2], and a Scala library would adopt [PVP][3]. (Apparently, using the first two numbers of the version, e.g. 1.2.x, to mean major version has a name, and it's Haskell Package Versioning Policy, or PVP for short).
+
+Unfortunately, the reality was different and while some Scala libraries use PVP (e.g., Play framework, the standard library, …), some others do use Semantic Versioning (e.g., `scala-collection-compat`, `cats`, …). This makes eviction warnings less reliable since many of these warnings are false positive, and, in practice, [they are ignored by developers][7].
+
+## Early SemVer and sbt-version-policy
+
+If you're maintaining a library, the first thing you could do to help is picking a version scheme. Our recommendation for libraries going forward is adopting **Early Semver** as described in the documentation page [Binary Compatibility for Library Authors][4]. Please read the linked post by Jacob Wang for more details on binary compatibility and source compatibility.
+
+Given a version number `major.minor.patch`, you MUST increment the:
+
+1. `major` version if backward **binary compatibility** is broken,
+2. `minor` version if backward **source compatibility** is broken, and
+3. `patch` version to signal neither binary nor source incompatibility.
+
+- When the `major` version is `0`, a minor version increment MAY contain **both source and binary breakages**, but a patch version increment MUST remain **binary compatible**.
+
+We call this Early SemVer, because according to the [Semantic Versioning Spec][2] there are no guarantees between any versions when the major version is `0`. In the Scala library ecosystem, though, we often start guaranteeing binary compatibility for `0.y.z` like sbt 0.13 and Scala.JS 0.6.
```

Comment:
"We call this" is ambiguous who "we" is. Is "early semver" standard terminology, or is it sbt-specific?

Reply:
early-semver is a version scheme Alex came up as a name in https://github.com/coursier/versions/issues/7, and I've been pushing it as implementation in Coursier, sbt, and elsewhere.